Maestro/src/Microsoft.DotNet.Maestro.WebApi/Controllers/CommitPushedController.cs

```diff
@@ -82,5 +96,82 @@ public class CommitPushedController : ApiController
 
             Trace.TraceInformation("CommitPushed Complete");
         }
+
+        private void VerifyPayloadSignature(byte[] requestContent)
+        {
+            string payloadSignature = GetPayloadSignature();
+            string computedHash = ComputeHashHexdigest(requestContent);
+
+            if (!CryptographicEquals(payloadSignature, computedHash))
+            {
+                Trace.TraceWarning($"Payload signature: {payloadSignature}");
+                Trace.TraceWarning($"Computed hash: {computedHash}");
+
+                throw new Exception("The payload signature specified by the X-Hub-Signature header does not match " +
+                    "the hash computed from the request.  This could indicate that the repo's Maestro webhook has " +
+                    "an incorrect Secret, the request came from an unauthorized source, or the request was corrupted.");
+            }
+        }
+
+        private string GetPayloadSignature()
+        {
+            IEnumerable<string> headerValues;
+            string requestXHubSignature = string.Empty;
+            if (Request.Headers.TryGetValues("X-Hub-Signature", out headerValues))
+            {
+                requestXHubSignature = headerValues.FirstOrDefault();
+            }
+
+            if (string.IsNullOrWhiteSpace(requestXHubSignature) || !requestXHubSignature.StartsWith("sha1="))
+            {
+                throw new Exception("The HTTP request does not contain a valid X-Hub-Signature header.  Verify that " +
+                    "the repo's Maestro webhook is configured with a Secret.");
+            }
+
+            return requestXHubSignature.Substring(5).ToUpper();
+        }
+
+        private string ComputeHashHexdigest(byte[] requestContent)
+        {
+            byte[] hmacKey = Encoding.UTF8.GetBytes(Config.Instance.WebhookSecretToken);
+
+            using (HMACSHA1 hmac = new HMACSHA1(hmacKey))
+            {
+                byte[] computedHash = hmac.ComputeHash(requestContent);
+                return BitConverter.ToString(computedHash).Replace("-", string.Empty);
+            }
+        }
+
+        // Test two strings for equality using a constant-time algorithm such that the execution time varies
+        // only on the length of the data, not the contents thereof, to prevent timing attacks. This method
+        // assumes that the strings are indexable.
```

Comment:
Something about combining characters, and non-normalized representations of data.  ASCII's all happy, it's things beyond char 127 that get a bit complex.